File: backend/core/startup_progress.py
# PIP Core - startup progress, for the app's launch screen.
#
# The Flutter client used to show a spinner and one of two sentences chosen by
# a RETRY COUNTER: "Starting PIP..." for the first seven attempts, "Still
# preparing things" after that. Neither was derived from anything actually
# happening - the app has no way to see inside a backend that is not answering
# yet, so after eight seconds it guessed, and said the same thing whether the
# database was being decrypted or nothing was running at all.
#
# WHY A FILE AND NOT AN ENDPOINT. The obvious design is GET /startup, and it
# cannot work: FastAPI's lifespan blocks serving until it finishes, so an
# endpoint cannot describe the work that is currently stopping it from
# existing. Nothing HTTP-shaped can report on a server that is not up.
#
# A file can, and this app already reads one on exactly this path - it polls
# data/api_token.txt in the same retry loop, for the same reason (the token is
# generated on first run and cannot be known at build time). So the launch
# screen is not being taught a new trick; it is reading a second file next to
# the one it already reads.
#
# APPEND-ONLY, one JSON object per line. A phase is a fact that happened, not a
# state to be overwritten, and appending is the one write that cannot leave a
# reader looking at a half-replaced file. It also means PowerShell and Python
# can both write here without coordinating - the launcher owns the phases
# before uvicorn exists, this module owns the ones after.
#
# NEVER RAISES. A launch screen is a courtesy. Taking down a startup because
# the courtesy could not be written would invert the priority exactly the way
# trace.stage_log() already declines to.

import json
import logging
import os
from pathlib import Path
from typing import Any

from backend.core.types import now_utc

logger = logging.getLogger(__name__)

DEFAULT_PROGRESS_PATH = Path(__file__).parent.parent.parent / "data" / "startup.jsonl"


def progress_path() -> Path:
    """Where phases are written. PIP_STARTUP_PROGRESS_PATH overrides, for tests."""
    override = os.environ.get("PIP_STARTUP_PROGRESS_PATH")
    return Path(override) if override else DEFAULT_PROGRESS_PATH


def report(phase: str, detail: str = "") -> None:
    """
    Records that [phase] has been reached.

    Ordering comes from the file, not from a timestamp: now_utc() has second
    resolution and a whole startup fits inside one second, which is the same
    trap trace.get_trace() documents ordering by id to avoid.
    """
    try:
        path = progress_path()
        path.parent.mkdir(parents=True, exist_ok=True)
        line = json.dumps({"phase": phase, "detail": detail, "at": now_utc()})
        with open(path, "a", encoding="utf-8") as handle:
            handle.write(line + "\n")
    except Exception as exc:  # pragma: no cover - defensive, see module docstring
        logger.debug(f"Could not record startup phase {phase!r}: {exc}")


def reset() -> None:
    """
    Clears the file for a fresh launch.

    Called by whoever starts the run - the launcher for a normal double-click,
    this module's own caller otherwise. Without it the previous launch's
    phases are still there and a launch screen would show a completed
    checklist before anything had happened. The client has a second guard for
    the same hazard (it ignores a file older than its own start), because a
    backend started by hand never goes through the launcher at all.
    """
    try:
        path = progress_path()
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text("", encoding="utf-8")
    except Exception as exc:  # pragma: no cover - defensive
        logger.debug(f"Could not reset startup progress: {exc}")


def read() -> list[dict[str, Any]]:
    """
    The phases recorded so far, in the order they happened.

    Skips any line that does not parse. A reader can arrive mid-write, and one
    torn line is not a reason to report nothing - the phases before it are
    still true.
    """
    try:
        raw = progress_path().read_text(encoding="utf-8")
    except OSError:
        return []

    phases: list[dict[str, Any]] = []
    for line in raw.splitlines():
        line = line.strip()
        if not line:
            continue
        try:
            entry = json.loads(line)
        except ValueError:
            continue
        if isinstance(entry, dict) and isinstance(entry.get("phase"), str):
            phases.append(entry)
    return phases

File: frontend/flutter/lib/startup_progress.dart
// What the launch screen shows instead of a spinner and a guess.
//
// The old screen picked between two sentences using a RETRY COUNTER: "Starting
// PIP..." for the first seven attempts, "Still preparing things - this can
// take a little longer on first launch..." after that. Neither came from
// anything happening. After eight seconds it said the same thing whether the
// database was being decrypted or nothing was running at all, which is the
// difference between "wait" and "this is broken".
//
// The real phases are written to data/startup.jsonl - by scripts/launch_pip.ps1
// for the parts that happen before uvicorn exists (Ollama, the password), and
// by backend/core/startup_progress.py from the instance lock onwards. A file
// rather than an endpoint because FastAPI's lifespan blocks serving until it
// finishes: nothing HTTP-shaped can report on a server that is not up yet.
// This app already polls a file on exactly this path for exactly this reason
// (data/api_token.txt), so it is reading a second file next to one it reads.
//
// Part 14.4 still holds. Nothing here decides anything: the backend says which
// phase it reached, and this turns a phase id into a sentence. An id it does
// not recognise is shown as itself rather than dropped, because an unfamiliar
// phase during a slow startup is exactly when someone needs to see it.

import 'dart:convert';

/// The phases a launch goes through, in order, with what to call each on
/// screen. Order is this list's, not the file's: the file records what
/// happened, and a launch that skipped a step should still show the step in
/// the place it would have occupied.
const startupPhaseLabels = <String, String>{
  'ollama': 'Local model service',
  'key': 'Unlocking your data',
  'backend': 'Starting PIP Core',
  'lock': 'Checking nothing else is running',
  'ready': 'Backend listening',
};

enum StartupPhaseState { done, current, pending }

class StartupPhase {
  final String id;
  final String label;
  final String detail;
  final StartupPhaseState state;
  const StartupPhase({
    required this.id,
    required this.label,
    required this.detail,
    required this.state,
  });
}

/// Parses the raw file into the ordered checklist to draw.
///
/// A phase earlier in the canonical order than the newest reported one is
/// shown complete even if its own line never arrived. That is not a guess
/// about the system: the launcher reports Ollama and the key on every path
/// where they are part of THIS launch, and omits them only when the backend
/// was already running - in which case they genuinely did happen, on the
/// launch that started it.
///
/// Returns an empty list for empty or unparseable input, which the caller
/// treats as "no information" rather than "nothing has happened".
List<StartupPhase> parseStartupProgress(String contents) {
  final reported = <String, String>{};
  final order = <String>[];

  for (final line in const LineSplitter().convert(contents)) {
    if (line.trim().isEmpty) continue;
    Object? decoded;
    try {
      decoded = jsonDecode(line);
    } catch (_) {
      // A reader can arrive mid-write. One torn line is not a reason to
      // report nothing - the phases before it are still true.
      continue;
    }
    if (decoded is! Map) continue;
    final id = decoded['phase'];
    if (id is! String || id.isEmpty) continue;
    final detail = '${decoded['detail'] ?? ''}';
    // A detail that merely restates the label is noise - "Backend listening /
    // backend listening" is how this first shipped. Dropped here rather than
    // only fixed at the writers, because there are three of them (this file's
    // labels, the PowerShell launcher, and the Python module) and nothing else
    // makes them agree.
    reported[id] =
        detail.toLowerCase() == (startupPhaseLabels[id] ?? '').toLowerCase() ? '' : detail;
    order.add(id);
  }

  if (order.isEmpty) return const [];

  final known = startupPhaseLabels.keys.toList();
  final latest = order.last;
  // An unknown phase id cannot be placed in the canonical order, so it is
  // treated as the furthest point reached and everything known is complete.
  final reachedIndex = known.indexOf(latest);

  final phases = <StartupPhase>[
    for (var i = 0; i < known.length; i++)
      StartupPhase(
        id: known[i],
        label: startupPhaseLabels[known[i]]!,
        detail: reported[known[i]] ?? '',
        state: reachedIndex < 0
            ? StartupPhaseState.done
            : i < reachedIndex
                ? StartupPhaseState.done
                : i == reachedIndex
                    ? StartupPhaseState.current
                    : StartupPhaseState.pending,
      ),
  ];

  // Anything the file reported that this build has no label for still gets a
  // row, under its own id. A phase added to the backend and not here is a
  // version skew, and the launch screen is the worst place to hide one.
  for (final id in order) {
    if (startupPhaseLabels.containsKey(id)) continue;
    phases.add(StartupPhase(
      id: id,
      label: id,
      detail: reported[id] ?? '',
      state: id == latest ? StartupPhaseState.current : StartupPhaseState.done,
    ));
  }

  return phases;
}

/// Whether the backend has reported that it is serving.
bool startupIsReady(List<StartupPhase> phases) =>
    phases.any((p) => p.id == 'ready' && p.state != StartupPhaseState.pending);
